common-workflow-language/cwlviewer
# Migrate from MongoDB to Postgres

Closes #254 

## Description


Replacing MongoDB by Postgres in CWLViewer.

- [x] update `pom.xml` dependencies, removing Mongo, adding Postgres + Hibernate (or data/jdbc)
- [x] replace the Spring configuration used by Mongo, probably add one for Postgres (pool settings, etc)
- [x] update the repositories to use Hibernate or JDBC instead of Data+Mongo
- [x] update entities and other annotations & classes as needed
- [x] add database migration, with initial DDL, but try to include some instructions for future DML's too
- [x] update tests (will probably need [TestContainers](https://github.com/testcontainers/testcontainers-java) to spin up a Postgres the QueuedWorkflowRepository tests)
  - [x] simple hello world workflow visualization
  - [x] cron jobs
- [x] update containers
- [x] send to review and write notes about anything important to observe when migrating, any docs, etc

## Motivation and Context


Licensing issues with MongoDB, see linked issue.

## How Has This Been Tested?




- [x] via unit tests
- [x] manually running the viewer
  * [x] running `mvn spring-boot:run` to trigger the flyway migrations
- [x] in GitHub Actions CI

## Screenshots (if appropriate):

## Types of changes


## Manual tests

- `WorkflowRepository`
  * [x] `findByRetrievedFrom` (tested by submitting the `compile.wdl` example Workflow in the landing page)
  * [x] `findByCommitAndPath` (tested by accessing `http://localhost:8080/git/767d700e602805112a4c953d166e570cddfa2605/workflows/compile/compile1.cwl?part=main&format=jsonld`)
  * [x] `findByCommit` (tested by accessing `http://localhost:8080/git/767d700e602805112a4c953d166e570cddfa2605/workflows/compile/compile1.cwl?format=raw`)
  * [x] `findAllByOrderByRetrievedOnDesc` (tested by clicking the "Explore" top menu link)
  * [x] `findByLabelContainingOrDocContainingIgnoreCase` (tested by searching for "aaa" and "compile" in the search bar, after submitting the `compile.cwl` workflow)
- `QueuedWorkflowRepositoryImpl`
  * [x] `findByRetrievedFrom` (tested in unit test `QueuedWorkflowRepositoryTest.java`)
  * [x] `deleteByTempRepresentation_RetrievedFrom` (tested in unit test `QueuedWorkflowRepositoryTest.java`)
- `QueuedWorkflowRepository`
  * [x] `deleteByTempRepresentation_RetrievedOnLessThanEqual` (tested by adding a queued workflow manually, below, and changing the scheduler to temporarily run every minute in `application.properties`)
  * [ ] `findByTempRepresentation_RetrievedOnLessThanEqual` (**not used??**)

To add a really old queued workflow:

```sql
insert into queued_workflow (id, cwltool_status, cwltool_version, message, temp_representation, workflow_list)
values ('bruno-testing', '{}'::jsonb, '1', 'yo', '{
"retrievedOn": "2016-02-04 00:45:56.000"
}'::jsonb, '{}'::jsonb);
```

Then query the table:

```sql
select * from queued_workflow;
```

And wait for the log line to confirm, or use a debugger. The log line should look like:

```
2022-02-24 22:44:23,001 INFO  [scheduling-1] org.commonwl.view.Scheduler: Deleting queued workflows older than or equal to Wed Feb 23 22:44:23 NZDT 2022
2022-02-24 22:44:23,004 INFO  [scheduling-1] org.commonwl.view.Scheduler: 1 Old queued workflows removed
```

## Checklist:
